File: 2022/06/part_one.go
package main

// go run 2022/06/part_one.go 2022/06/input.txt

import (
	"bufio"
	"log"
	"os"
)

func findMarkerIndex(s string) int {
	var buffer [4]rune
	isUnique := func() bool {
		set := map[rune]struct{}{}
		for _, r := range buffer {
			_, ok := set[r]
			if ok {
				return false
			}
			set[r] = struct{}{}
		}

		return true
	}

	for i, r := range s {
		buffer[i%len(buffer)] = r

		log.Println(r, buffer)

		if i >= 4 && isUnique() {
			// The character at index i is the i+1 character
			return i + 1
		}
	}

	return -1
}

func main() {
	fileName := os.Args[1]

	f, err := os.Open(fileName)
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := scanner.Text()
		marker := findMarkerIndex(line)
		log.Println("marker: ", marker)
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}
}

File: 2022/06/partOne.scala
package advent.of.code
// scala 2022/06/partOne.scala 2022/06/input.txt

import scala.annotation.tailrec
import scala.io.Source

@tailrec
def findMarkerIndex(data: String, idx: Int = 0, buffer: String = "", bufferSize: Int = 4): Int = {
    def isUnique(b: String): Boolean =
        b.length == b.distinct.length

    if (data.nonEmpty) then
        val ch = data.take(1)
        val newData = data.substring(1)
        val b = ch.toString + buffer.take(bufferSize - 1)

        if (idx >= bufferSize) then
            if (isUnique(b)) then
                idx + 1
            else
                findMarkerIndex(newData, idx + 1, b, bufferSize)
        else
            findMarkerIndex(newData, idx + 1, b, bufferSize)
    else
        -1
}

@main def template(fileName: String) = {
    val lines = Source.fromFile(fileName).getLines
    .map(line => findMarkerIndex(line))
    .foreach(println)
}
